Better --port & --bridge-port options validation in CLI

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#368 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have improved validations for `--port` and `--bridge` options to require a value within reasonable bounds.

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

jonathanknowles committed:
Write version string to the log when starting the wallet server.

KtorZ committed:
constraints 'Port' type between valid bounds

We use to accept any 'Int' as ports on the CLI, which causes only
a late error when trying to open a socket on an invalid TCP port.

KtorZ committed:
re-generate nix machinery

piotr-iohk commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -104,11 +113,32 @@ import qualified Data.Text.IO as TIO
+-- and only "bindable" through root privileges.
+--
+-- Ports above 49151 are also reserved for ephemeral connections. This
+-- leaves us with a valid range of [[1024;49151]] for TCP ports.
```
:+1:

piotr-iohk commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Port.hs:
```diff
@@ -51,6 +54,30 @@ import Test.Integration.Framework.DSL
+                , ("launch", "--port", getPort maxBound + 14)
+                , ("launch", "--bridge-port", getPort minBound - 14)
+                , ("launch", "--bridge-port", getPort maxBound + 14)
+                ]
```
That's nice (btw. any reason for `14`?). I'd also add more negative and positive cases for some values that usually tend to be _suspicious_:
 - boundary values (here `minBound` and `maxBound` and their nearest vicinity (`_Bound + 1`, `_Bound - 1`))
 - arbitrary string
 - negative value
 - `0`
 - ...

KtorZ replied:
No reason actually, it's completely arbitrary. I'd suggest that we could add more tests like this at the `unit` level, and not in integration tests, as they would be less costly to run. The above make sure that the parser gets called, and we can extensively test the parser in unit tests, what do you think ?

KtorZ replied:
See: https://github.com/input-output-hk/cardano-wallet/pull/425/files#diff-7e848cbd8fd36a221b008a73bd1e550aR55

KtorZ replied:
Also, I've made the above more arbitrary to make it clear that there's nothing special about them.

piotr-iohk replied:
That's ok, just want to point out that we had some cases previously where parser that worked fine on `unit` level didn't work as expected at `integration` level (address-pool-gap = -1000 was 24). Therefore, in principle, I'd prefer having few more _classic_ black box cases like that on integration level as well. Ofc the risk here maybe unlikely, and potential gain from lest costly test is better (on the other hand who would've known that -1000 is 24 ¯\_(ツ)_/¯ ;) )

KtorZ replied:
Well... the parser wasn't actually fine on the unit level either in that case :grimacing: ... because the same code is used in both the API and the primitive types. Generally speaking however, I do agree about the approach on integration tests. Yet here, I think we've a nice trade-off -> we know the same parser is being used (cf error message, I know it's a poor guarantee but well), and we have it well-tested at the unit-level.

jonathanknowles replied:
If these numbers are all arbitrarily chosen, then I would suggest adding a comment to that effect. For example:
```hs
        let tests =
                -- The numbers below are all chosen arbitrarily:
                [ ("serve", "--port", getPort minBound - 14)
                , ("serve", "--port", getPort maxBound + 42)
                , ("serve", "--bridge-port", getPort minBound - 6)
                , ("serve", "--bridge-port", getPort maxBound + 523)
                , ("launch", "--port", getPort minBound - 1)
                , ("launch", "--port", getPort maxBound + 59375)
                , ("launch", "--bridge-port", getPort minBound - 5621)
                , ("launch", "--bridge-port", getPort maxBound + 1)
                ]
```

KtorZ committed:
Make port numbers a bit more arbitrary in tests

This to show that the value has no particular meaning. Property tests would also convey that better

KtorZ committed:
add more negative unit test cases for port decoding

piotr-iohk reviewed (approved):
lgtm :+1:

jonathanknowles committed:
Pass CLI argument `--min-log-severity` through when calling `serve` from `launch`.

KtorZ committed:
Merge branch 'jonathanknowles/wallet-server-startup-log-entry' into KtorZ/--port-validation

KtorZ committed:
re-generate nix machinery
